[Bug] ChibiOS compilation error: '__builtin_memcpy' offset [0, 21] is out of the bounds [0, 0]

I'm not sure whether I should open this issue here, on tmk, or chibios, but it's not in any of the submodules, so I guess I'll start here. If it doesn't belong here, please redirect me to the correct place.

## Describe the Bug

When trying to compile the firmware for ergodox_infinity, I get the following output:

```
$ MAKEFLAGS=-j1 make ergodox_infinity:default MASTER=right

QMK Firmware 0.12.38
Making ergodox_infinity with keymap default

arm-none-eabi-gcc (Arch Repository) 11.1.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Compiling: keyboards/ergodox_infinity/keymaps/default/keymap.c                                      [OK]
Compiling: quantum/command.c                                                                        [OK]
Compiling: tmk_core/common/chibios/bootloader.c                                                    tmk_core/common/chibios/bootloader.c: In function 'bootloader_jump':
tmk_core/common/chibios/bootloader.c:90:5: error: '__builtin_memcpy' offset [0, 21] is out of the bounds [0, 0] [-Werror=array-bounds]
   90 |     __builtin_memcpy((void *)VBAT, (const void *)sys_reset_to_loader_magic, sizeof(sys_reset_to_loader_magic));
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
 [ERRORS]
 |
 |
 |
make[1]: *** [tmk_core/rules.mk:417: .build/obj_ergodox_infinity_default/common/chibios/bootloader.o] Error 1
Make finished with errors
make: *** [Makefile:523: ergodox_infinity:default] Error 1
```

This is after a second make run, to reduce the output for files that have already previously been compiled successfully. 

## System Information

 - Keyboard:  Ergodox Infinity
 - Operating system: ArchLinux
 - ARM GCC version: 11.1.0
 - QMK Firmware version: 0.12.38 (I'm currently on commit hash a5d7e83985ae07689cc71ab1dd2755a8ac335df3)
 - Any keyboard related software installed? None

## Additional Context

A few days ago, this compilation ran just fine, and this error happened both before rebasing on the latest master commid and after. My guess is that something changed in my local environment, since running it in the Docker container from the provided Dockerfile, works flawlessly. 

The `arm-none-eabi-gcc` package on archlinux received an update 4 days ago, so I assume that that is the culprit, I'm just not sure why.

I have set up a Dockerfile that reproduces the error (at least at time of writing) using the archlinux base image:

```Dockerfile
FROM archlinux

RUN pacman -Syy \
	&& pacman -S --noconfirm \
		make \
		git \
		diffutils \
		arm-none-eabi-gcc \
		qmk \
	&& git clone https://github.com/qmk/qmk_firmware

WORKDIR /qmk_firmware
RUN git checkout a5d7e83985ae07689cc71ab1dd2755a8ac335df3 \
	&& make git-submodule

CMD make ergodox_infinity:default
```

Same thing :/

```
Compiling: tmk_core/common/chibios/bootloader.c                                                    tmk_core/common/chibios/bootloader.c: In function 'bootloader_jump':
tmk_core/common/chibios/bootloader.c:90:5: error: '__builtin_memcpy' offset [0, 21] is out of the bounds [0, 0] [-Werror=array-bounds]
   90 |     __builtin_memcpy((void *)VBAT_BASE, (const void *)sys_reset_to_loader_magic, sizeof(sys_reset_to_loader_magic));
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      ```

--- a/tmk_core/common/chibios/bootloader.c
+++ b/tmk_core/common/chibios/bootloader.c
@@ -103,7 +103,8 @@ void enter_bootloader_mode_if_requested(void) {
 #        define SCB_AIRCR_VECTKEY_WRITEMAGIC 0x05FA0000
 const uint8_t              sys_reset_to_loader_magic[] = "\xff\x00\x7fRESET TO LOADER\x7f\x00\xff";
 __attribute__((weak)) void bootloader_jump(void) {
-    __builtin_memcpy((void *)VBAT, (const void *)sys_reset_to_loader_magic, sizeof(sys_reset_to_loader_magic));
+    void *volatile vbat = (void *)VBAT;
+    __builtin_memcpy(vbat, (const void *)sys_reset_to_loader_magic, sizeof(sys_reset_to_loader_magic));
     // request reset
     SCB->AIRCR = SCB_AIRCR_VECTKEY_WRITEMAGIC | SCB_AIRCR_SYSRESETREQ_Msk;
 }

